redundant type cast causes errors with VCC

### Description

This can be triggered when using `asyncdispatch`:

```nim
import asyncdispatch

proc bug() {.async} =
    echo "bug"

proc main() =
    bug().waitFor()

main()
```

Part of generated C code:

```c
{
  tyProc__HzVCwACFYM9cx9aV62PdjtuA T30_;
  nimZeroMem((void*)(&T30_), sizeof(tyProc__HzVCwACFYM9cx9aV62PdjtuA));
  T30_.ClP_0 = bugNimAsyncContinue__bug_u5; T30_.ClE_0 = colonenvP_;
  addCallback__pureZasyncfutures_u676(nextX60gensym4_, ((tyProc__HzVCwACFYM9cx9aV62PdjtuA) (T30_)));
  if (NIM_UNLIKELY(*nimErr_)) goto LA7_;
}
```

Converting `T30_` to `tyProc__HzVCwACFYM9cx9aV62PdjtuA` is redundant, since it is just a `tyProc__HzVCwACFYM9cx9aV62PdjtuA`. GCC allows such redundant type conversions, while VCC rejects them.

### Nim Version

```
Nim Compiler Version 2.2.0 [Windows: amd64]
Compiled at 2024-10-02
Copyright (c) 2006-2024 by Andreas Rumpf

active boot switches: -d:release
```

### Current Output


When built with VCC, it reports:

```
error C2440: 'type cast': cannot convert from 'tyProc__HzVCwACFYM9cx9aV62PdjtuA' to 'tyProc__HzVCwACFYM9cx9aV62PdjtuA'
```

### Expected Output

```text
The `bug.nim` should compile with VCC.
```


### Known Workarounds

Use GCC.

### Additional Information

_No response_

--- a/compiler/ccgexprs.nim
+++ b/compiler/ccgexprs.nim
@@ -2520,6 +2520,8 @@ proc genSomeCast(p: BProc, e: PNode, d: var TLoc) =
       putIntoDest(p, d, e, cDeref(cCast(ptrType(destTyp), wrapPar(cAddr(val)))), a.storage)
     elif etyp.kind == tyBool and srcTyp.kind in IntegralTypes:
       putIntoDest(p, d, e, cOp(NotEqual, rdCharLoc(a), cIntValue(0)), a.storage)
+    elif etyp.kind == tyProc and srcTyp.kind == tyProc and sameBackendType(etyp, srcTyp):
+      expr(p, e[1], d)
     else:
       if etyp.kind == tyPtr:
         # generates the definition of structs for casts like cast[ptr object](addr x)[]

--- a/compiler/vtables.nim
+++ b/compiler/vtables.nim
@@ -15,7 +15,6 @@ proc dispatch(x: Base, params: ...) =
 
   var disp = newNodeI(nkIfStmt, base.info)
 
-  var vTableAccess = newNodeIT(nkBracketExpr, base.info, base.typ)
   let nimGetVTableSym = getCompilerProc(g, "nimGetVTable")
   let ptrPNimType = nimGetVTableSym.typ.n[1].sym.typ
 
@@ -33,7 +32,7 @@ proc dispatch(x: Base, params: ...) =
     dispatchObject,
     newIntNode(nkIntLit, index)
   )
-  getVTableCall.typ() = base.typ
+  getVTableCall.typ() = getSysType(g, unknownLineInfo, tyPointer)
   var vTableCall = newNodeIT(nkCall, base.info, base.typ.returnType)
   var castNode = newTree(nkCast,
         newNodeIT(nkType, base.info, base.typ),

